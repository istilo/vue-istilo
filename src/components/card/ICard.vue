<script setup lang="ts">
import { computed, useSlots } from 'vue';

const slots = useSlots();

const props = withDefaults(
	defineProps<{
		/**
		 * Add a background color
		 *
		 * @values default, primary, secondary
		 */
		color?: string;

		/**
		 * Adding box-shadow when mouse over on the card
		 */
		isHover?: boolean;

		/**
		 * Change the card size by adjusting the gutters
		 *
		 * @values small, large
		 */
		size?: string;
	}>(),
	{
		color: 'default',
		hover: false,
	}
);

const getClasses = computed(() => {
	return [
		'i-card',
		props.color ? `i-card-${props.color}` : '',
		props.isHover ? 'i-card-hover' : '',
		props.size ? `i-card-${props.size}` : '',
	];
});
</script>

<template>
	<div :class="getClasses">
		<div
			v-if="slots.header"
			class="i-card-header">
			<slot name="header" />
		</div>
		<div class="i-card-body">
			<slot />
		</div>
		<div
			v-if="slots.footer"
			class="i-card-footer">
			<slot name="footer" />
		</div>
	</div>
</template>
