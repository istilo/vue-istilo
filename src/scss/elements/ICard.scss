@use 'sass:math';

@import '@/scss/helpers';
@import '@/scss/functions';
@import '@/scss/mixins';

// ============================================================
// Name:            Card
// Description:     Component to create boxed content containers
//
// Component:       `#{$prefix}card`
//
// Sub-objects:     `#{$prefix}card-body`
//                  `#{$prefix}card-header`
//                  `#{$prefix}card-footer`
//                  `#{$prefix}media-*`
//                  `#{$prefix}card-title`
//                  `#{$prefix}card-badge`
//
// Modifiers:       `#{$prefix}card-hover`
//                  `#{$prefix}card-default`
//                  `#{$prefix}card-primary`
//                  `#{$prefix}card-secondary`
//                  `#{$prefix}card-small`
//                  `#{$prefix}card-large`
//
// Uses:            `#{$prefix}grid-stack`
//
// ============================================================

// Variables
// ============================================================

$card-border-color: border(color) !default;
$card-border-width: border(width) !default;
$card-border-radius: 10px !default;

$card-body-padding-horizontal: gutter(small) !default;
$card-body-padding-vertical: gutter(small) !default;

$card-body-padding-horizontal-l: gutter(medium) !default;
$card-body-padding-vertical-l: gutter(medium) !default;

$card-header-padding-horizontal: gutter(small) !default;
$card-header-padding-vertical: math.div(gutter(small), 2) !default;

$card-header-padding-horizontal-l: gutter(medium) !default;
$card-header-padding-vertical-l: math.div(gutter(medium), 2) !default;

$card-footer-padding-horizontal: gutter(small) !default;
$card-footer-padding-vertical: gutter(small) !default;

$card-footer-padding-horizontal-l: gutter(medium) !default;
$card-footer-padding-vertical-l: gutter(medium) !default;

$card-title-font-size: font-size(large) !default;
$card-title-line-height: 1.4 !default;

$card-badge-top: gutter(small) !default;
$card-badge-right: gutter(small) !default;

$card-hover-background: theme-color(default) !default;

$card-default-background: $global-background !default;
$card-default-color: color(default) !default;
$card-default-title-color: color(emphasis) !default;
$card-default-hover-background: darken($card-default-background, 5%) !default;

$card-primary-background: theme-color(primary) !default;
$card-primary-color: color(inverse) !default;
$card-primary-title-color: theme-color(primary) !default;
$card-primary-hover-background: darken($card-primary-background, 5%) !default;

$card-secondary-background: theme-color(secondary) !default;
$card-secondary-color: color(inverse) !default;
$card-secondary-title-color: color(inverse) !default;
$card-secondary-hover-background: darken($card-secondary-background, 5%) !default;

$card-small-body-padding-horizontal: gutter(xsmall) !default;
$card-small-body-padding-vertical: gutter(xsmall) !default;
$card-small-header-padding-horizontal: gutter(xsmall) !default;
$card-small-header-padding-vertical: gutter(xsmall) !default;
$card-small-footer-padding-horizontal: gutter(xsmall) !default;
$card-small-footer-padding-vertical: gutter(xsmall) !default;

$card-large-body-padding-horizontal-l: gutter(large) !default;
$card-large-body-padding-vertical-l: gutter(large) !default;
$card-large-header-padding-horizontal-l: gutter(large) !default;
$card-large-header-padding-vertical-l: math.div(gutter(large), 2) !default;
$card-large-footer-padding-horizontal-l: gutter(large) !default;
$card-large-footer-padding-vertical-l: math.div(gutter(large), 2) !default;

/* ============================================================
	Component: Card
   ============================================================ */

.#{$prefix}card {
	position: relative;
	box-sizing: border-box;

	border-radius: $card-border-radius;
	transition: all 0.3s ease;
}

/* Sections
   ============================================================ */

.#{$prefix}card-body {
	padding: $card-body-padding-vertical $card-body-padding-horizontal;
}

.#{$prefix}card-header {
	padding: $card-header-padding-vertical $card-header-padding-horizontal;
	border-bottom: 1px solid $card-border-color;
}

.#{$prefix}card-footer {
	background-color: #fafafa;
	padding: $card-footer-padding-vertical $card-footer-padding-horizontal;
	border-top: 1px solid $card-border-color;
}

.#{$prefix}card-footer.#{$prefix}card-footer-button {
	padding: 0px;
}

.#{$prefix}card-footer-button > button {
	width: 100%;
	border-top-left-radius: 0px;
	border-top-right-radius: 0px;
}

/* Desktop and bigget */
@include respond-to(large) {
	.#{$prefix}card-body {
		padding: $card-body-padding-vertical-l $card-body-padding-horizontal-l;
	}

	.#{$prefix}card-header {
		padding: $card-header-padding-vertical-l $card-header-padding-horizontal-l;
	}

	.#{$prefix}card-footer {
		padding: $card-footer-padding-vertical-l $card-footer-padding-horizontal-l;
	}
}

/*
 * Micro clearfix
 */

.#{$prefix}card-body,
.#{$prefix}card-header,
.#{$prefix}card-footer {
	@include clearfix;
}

/*
 * Remove margin from the last child
 */

.#{$prefix}card-body > :last-child,
.#{$prefix}card-header > :last-child,
.#{$prefix}card-footer > :last-child {
	margin-bottom: 0;
}

/* Media
   ============================================================ */

/*
 * 1. Set a specific dimension
 * 2. Set overflow hidden
 */
[class*='#{$prefix}card-media-'] {
	/* 1 */
	height: 400px;
	width: 100%;
	overflow: hidden;
}

/*
 * 1. Set the dimensions
 * 2. Make the object fit the parent
 * We set the object to cover and overflow hidden on its parent to avoid
 * inconsistent sizes of images, it will mask the image of its parent container
 * which are "#{$prefix}card-media-*"
 */
[class*='#{$prefix}card-media-'] img,
[class*='#{$prefix}card-media-'] video {
	/* 1 */
	height: 100%;
	width: 100%;

	/* 2 */
	object-fit: cover;
	object-position: center;
}

/* Title
   ============================================================ */

.#{$prefix}card-title {
	font-size: $card-title-font-size;
	line-height: $card-title-line-height;
	margin-top: 0;
}

/* Dropdown
   ============================================================ */

.#{$prefix}card-dropdown {
	position: absolute !important;
	top: 15px !important;
	right: 25px !important;
}

/* Badge
   ============================================================ */

.#{$prefix}card-badge {
	position: absolute;
	top: $card-badge-top;
	right: $card-badge-right;
	z-index: 1;
}

.#{$prefix}card-badge-left {
	right: auto;
	left: $card-badge-right;
}

/*
 * Remove margin from adjacent element
 */

.#{$prefix}card-badge:first-child + * {
	margin-top: 0;
}

/* Hover modifier
   ============================================================ */

.#{$prefix}card-hover:not(.#{$prefix}card-default):not(.#{$prefix}card-primary):not(
		.#{$prefix}card-secondary
	):hover {
	box-shadow: box-shadow(large);
}

/* Style modifiers
   ============================================================ */

/*
 * Default
 */
.#{$prefix}card-default {
	background-color: $card-default-background;
	color: $card-default-color;

	border: $card-border-width solid $card-border-color;
}

.#{$prefix}card-default .#{$prefix}card-title {
	color: $card-default-color;
}

.#{$prefix}card-default.#{$prefix}card-hover:hover {
	box-shadow: box-shadow(large);
}

/*
 * Primary
 */
.#{$prefix}card-primary {
	background-color: $card-primary-background;
	color: $card-primary-color;
}

.#{$prefix}card-primary .#{$prefix}card-title {
	color: $card-primary-color;
}

.#{$prefix}card-primary.#{$prefix}card-hover:hover {
	box-shadow: 0px 30px 40px rgba(44, 52, 72, 0.3);
}

/*
 * Secondary
 */
.#{$prefix}card-secondary {
	background-color: $card-secondary-background;
	color: $card-secondary-color;
}

.#{$prefix}card-secondary .#{$prefix}card-title {
	color: $card-secondary-color;
}

.#{$prefix}card-secondary.#{$prefix}card-hover:hover {
	box-shadow: box-shadow(large);
}

/* Size modifiers
   ============================================================ */

.#{$prefix}card-small.#{$prefix}card-body,
.#{$prefix}card-small .#{$prefix}card-body {
	padding: $card-small-body-padding-vertical $card-small-body-padding-horizontal;
}

.#{$prefix}card-small .#{$prefix}card-header {
	padding: $card-small-header-padding-vertical $card-small-header-padding-horizontal;
}

.#{$prefix}card-small .#{$prefix}card-footer {
	padding: $card-small-footer-padding-vertical $card-small-footer-padding-horizontal;
}

/*
 * Large
 */

/* Desktop and bigger */
@include respond-to(large) {
	.#{$prefix}card-large.#{$prefix}card-body,
	.#{$prefix}card-large .#{$prefix}card-body {
		padding: $card-large-body-padding-vertical-l $card-large-body-padding-horizontal-l;
	}

	.#{$prefix}card-large .#{$prefix}card-header {
		padding: $card-large-header-padding-vertical-l $card-large-header-padding-horizontal-l;
	}

	.#{$prefix}card-large .#{$prefix}card-footer {
		padding: $card-large-footer-padding-vertical-l $card-large-footer-padding-horizontal-l;
	}
}
