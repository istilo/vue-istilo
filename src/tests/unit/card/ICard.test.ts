import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { shallowMount, VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import ICard from '@/components/card/ICard.vue';

enableAutoUnmount(afterEach);

describe('Istilo Card', () => {
	let wrapper: VueWrapper<any>;

	beforeEach(() => {
		wrapper = shallowMount(ICard, {
			propsData: {
				label: 'AV',
			},
			slots: {
				default: '<p>Card</p>',
			},
		});

		expect(wrapper.exists()).toBe(true);
	});

	afterEach(() => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('Should set a default class and children', () => {
		expect(wrapper.classes()).toContain('i-card');
		expect(wrapper.get('.i-card-body')).toBeDefined();
	});

	it('Should set a color modifier class name', async () => {
		await wrapper.setProps({
			color: 'primary',
		});

		expect(wrapper.classes()).toContain('i-card-primary');
	});

	it('Should set a hover modifier class name', async () => {
		await wrapper.setProps({
			isHover: true,
		});

		expect(wrapper.classes()).toContain('i-card-hover');
	});

	it('Should set a size modifier class name', async () => {
		await wrapper.setProps({
			size: 'large',
		});

		expect(wrapper.classes()).toContain('i-card-large');
	});

	it('Should set a size modifier class name', async () => {
		wrapper = shallowMount(ICard, {
			propsData: {
				label: 'AV',
			},
			slots: {
				default: '<p>Card</p>',
				header: `
					<template v-slot:header>
						<p>Header</p>
					</template>`,
				footer: `
					<template v-slot:footer>
						<p>Footer</p>
					</template>`,
			},
		});

		expect(wrapper.get('.i-card-body')).toBeDefined();
		expect(wrapper.get('.i-card-header')).toBeDefined();
		expect(wrapper.get('.i-card-footer')).toBeDefined();
	});
});
