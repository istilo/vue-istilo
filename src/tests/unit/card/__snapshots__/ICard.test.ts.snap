// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Istilo Card > Should set a color modifier class name 1`] = `
"<div class=\\"i-card i-card-primary\\" label=\\"AV\\">
  <!--v-if-->
  <div class=\\"i-card-body\\">
    <p>Card</p>
  </div>
  <!--v-if-->
</div>"
`;

exports[`Istilo Card > Should set a default class and children 1`] = `
"<div class=\\"i-card i-card-default\\" label=\\"AV\\">
  <!--v-if-->
  <div class=\\"i-card-body\\">
    <p>Card</p>
  </div>
  <!--v-if-->
</div>"
`;

exports[`Istilo Card > Should set a hover modifier class name 1`] = `
"<div class=\\"i-card i-card-default i-card-hover\\" label=\\"AV\\">
  <!--v-if-->
  <div class=\\"i-card-body\\">
    <p>Card</p>
  </div>
  <!--v-if-->
</div>"
`;

exports[`Istilo Card > Should set a size modifier class name 1`] = `
"<div class=\\"i-card i-card-default i-card-large\\" label=\\"AV\\">
  <!--v-if-->
  <div class=\\"i-card-body\\">
    <p>Card</p>
  </div>
  <!--v-if-->
</div>"
`;

exports[`Istilo Card > Should set a size modifier class name 2`] = `
"<div class=\\"i-card i-card-default\\" label=\\"AV\\">
  <div class=\\"i-card-header\\">
    <p>Header</p>
  </div>
  <div class=\\"i-card-body\\">
    <p>Card</p>
  </div>
  <div class=\\"i-card-footer\\">
    <p>Footer</p>
  </div>
</div>"
`;
